text/0000-inherent-traits.md

```diff
@@ -0,0 +1,91 @@
+- Feature Name: inherent_traits
+- Start Date: 2018-01-21
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Provides a mechanism to declare "inherent traits" for a type defined in the same crate. Methods on these traits are callable on instances
+of the specified type without needing to import the trait.
+
+# Motivation
+[motivation]: #motivation
+
+There are two similar cases where this is valuable:
+
+- Mapping object-oriented APIs.
+
+  When mapping these APIs to rust, base classes are usually mapped to traits: methods on those base classes will need to be callable on any
+  derived type. This is sub-optimal because to use a class method a user must now know which class in the hierachy defined that
+  method, so that they can import and use the corresponding trait. This knowledge is not required when using the same API from an
+  object-oriented language.
+
+- Frequently used types.
+
+  Sometimes getting the right abstractions require breaking up a type's implementation into many traits, with only a few methods per
+  trait. Every use of such a type results in a large number of imports to ensure the correct traits are in scope. If such a type is used
+  frequently, then this burden quickly becomes a pain point for users of the API.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The feature is implemented using a new attribute which can be applied to `impl` blocks:
+
+```rust
+pub struct Foo;
+
+trait Bar {
+    fn bar(&self);
+}
+
+impl Bar for Foo {
+    fn bar(&self) { println!("foo::bar"); }
+}
+
+#[include(Bar)]
+impl Foo {
+    fn foo(&self) { println!("foo::foo"); }
+}
+```
+
+The method `bar` is now callable on any instance of `Foo`, regardless of whether the `Bar` trait is currently in scope, or even whether
+the `Bar` trait is publically visible.
```

Comment:
Contemplating this for myself, what are the other options?

Some bad ones:
- Inherit `pub` from the trait: gets into all the "what's pub, really" discussions
- Put something on the methods: confusing to have `pub` there suddenly, doesn't work for trait methods with default impls
- Put something on the trait: no, traits shouldn't ever care about this

 A possibility: `#[inherent(A)]` puts `pub(A)` on them.  So just `#[inherent]` is `pub`, but you can use `#[inherent(crate)]` (or maybe `#[inherent(in foo::bar)]`?) to limit it if helpful.

But really, in the same crate the trait import isn't a big deal, so that's probably not worth bothering with, and just "it's always `pub`" (as the RFC says) is probably the way to go :+1: